pkg/model/bootstrapscript.go

```diff
@@ -131,7 +131,7 @@ func (b *BootstrapScript) buildEnvironmentVariables(cluster *kops.Cluster) (map[
 		env["GOSSIP_DNS_CONN_LIMIT"] = os.Getenv("GOSSIP_DNS_CONN_LIMIT")
 	}
 
-	if os.Getenv("S3_ENDPOINT") != "" {
+	if os.Getenv("S3_ENDPOINT") != "" && (!model.UseKopsControllerForNodeBootstrap(cluster) || b.ig.IsMaster()) {
```

Comment:
Given there's still IAM for `/addons/*` and `/secrets/dockerconfig`, is this not still needed?

Reply:
`/addons/*` could also be removed on worker nodes, as Protokube only applies them from the control plane.
The only remaining nit is `/secrets/dockerconfig`. Maybe we could keep this around if at least a `dockerconfig` secret exists?

Reply:
I guess a followup would be to also serve the secrets from kops-controller.

Reply:
Yes, that would be a good followup.

Reply:
Do we need to set these env vars in order for nodeup to be able to read dockerconfig?

Reply:
Only for non-AWS clouds that don't have object storage, like Hetzner.

Reply:
It looks like this would affect Hetzner and GCE using a custom S3 state stores and also using a Docker secret.

I think this change should be deferred until after the secrets are served from kops-controller.

Reply:
From security point of view, giving access to S3 fully is much worse than not having a Docker secret.

Reply:
Perhaps add a condition testing whether or not a dockerconfig secret exists?

Reply:
Using config for bootstrap switches to using `configserverSecretStore` automatically, which is not implemented.
https://github.com/kubernetes/kops/blob/0be79b25b74e5bec3ddd0907918605b39600e1dd/pkg/configserver/secretstore.go#L38-L41
We get away with it due to error not being checked:
https://github.com/kubernetes/kops/blob/0be79b25b74e5bec3ddd0907918605b39600e1dd/nodeup/pkg/model/secrets.go#L52
I would say that, unless I am missing something, this doesn't work in general.

Reply:
Please see https://github.com/kubernetes/kops/pull/14501/commits/8ba1fa46c3f36a5a9334ee52bbec085354e4446b for a quick implementation of serving secrets from kops-controller.